### PR TITLE
refactor: Token approval actions

### PR DESCRIPTION
**Pull request by garethfuller**
# Description

It shouldn't be possible to get to the last step of BalActionSteps if you don't approve the correct amounts. This PR adds a `postActionValidation` option to BalActionSteps which allows us to validate that the correct approval amount has been confirmed in the transaction flows. If the user accidentally approves a lower amount than required it doesn't progress the BalActionSteps and displays a warning that informs them that the correct approval amount was not confirmed. They can then re-submit the approval transaction until this `postActionValidation` is successful.

A large number of Fatal errors in Sentry seem to be to do with this issue of users getting to the last step in the BalActionSteps with incorrect approvals, so hopefully this significantly reduces the errors labeled "cannot estimate gas; transaction may fail or may require manual gas limit".

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Code refactor / cleanup

## How should this be tested?

- [ ] Pool creations approvals
- [ ] Staking approvals
- [ ] veBAL locking approvals

## Visual context

https://www.loom.com/share/684749bee3644064b3fee5f1eddb1201

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have requested at least 1 review (If the PR is significant enough, use best judgement here)
- [x] I have commented my code where relevant, particularly in hard-to-understand areas
- [x] If package-lock.json has changes, it was intentional.
- [x] The base of this PR is `master` if hotfix, `develop` if not

### CODE REVIEW

**Review by timjrobinson — APPROVED** [review]

**Comment by timjrobinson** [comment]
Tested on polygon and it worked correctly for me. Some minor code style issues, otherwise LGTM.

File: src/composables/approvals/useTokenApprovalActions.ts [hunk 1]
```diff
@@ -66,54 +69,152 @@ export default function useTokenApprovalActions(
...
+        return t('transactionSummary.approveForInvesting', [
+          getToken(address)?.symbol,
+        ]);
+    }
+  }
+
+  async function getApprovalsRequired(
+    amountsToApprove: AmountToApprove[],
+    spender: string
+  ): Promise<string[]> {
```
[the remark this answers is not in the document]
Comment (garethfuller) [reply src/composables/approvals/useTokenApprovalActions.ts]: Changed to return a subset of AmountToApprove[].

File: src/composables/approvals/useTokenApprovalActions.ts [hunk 2]
```diff
@@ -66,54 +69,152 @@ export default function useTokenApprovalActions(
...
+    await refetchAllowances();
+    const tokenAddresses = amountsToApprove.map(ata => ata.address);
+    const amounts = amountsToApprove.map(ata => ata.amount);
+    return approvalsRequired(tokenAddresses, amounts, spender);
```
[the remark this answers is not in the document]
Comment (garethfuller) [reply src/composables/approvals/useTokenApprovalActions.ts]: Refactored tokens provider to do as suggested.

File: src/composables/approvals/useTokenApprovalActions.ts [hunk 3]
```diff
@@ -1,50 +1,53 @@
...
-export default function useTokenApprovalActions(
-  tokenAddresses: Ref<string[]>,
-  amounts: Ref<string[]>,
-  actionType: ApprovalAction = ApprovalAction.AddLiquidity
-) {
+interface Params {
+  amountsToApprove: AmountToApprove[];
+  spender: string;
+  actionType: ApprovalAction;
+  forceMaxApprovals?: boolean;
```
[the remark this answers is not in the document]
Comment (garethfuller) [reply src/composables/approvals/useTokenApprovalActions.ts]: Done.

**Pull request merged**